build_tools/github_actions/runner/gcp/create_image.sh

```diff
@@ -26,10 +26,13 @@ ZONE="${ZONE:-us-central1-a}"
 PROJECT=iree-oss
 BASE_IMAGE="${BASE_IMAGE:-projects/ubuntu-os-cloud/global/images/ubuntu-2204-jammy-v20230531}"
 
+
 # We create the image using n1 machines with attached T4 GPUs. This image works
 # for the A100 machines as well though.
 GPU_MACHINE_TYPE="n1-standard-16"
-CPU_MACHINE_TYPE="e2-medium"
+# Temporary to try things out on GCP. Assume we should parameterize this script
+# eventually.
+CPU_MACHINE_TYPE="t2a-standard"
```

Comment:
Just a temporary thing so I can try this out on GCP.

Reply:
Yeah I think we'd just call it CPU_X86_64 and CPU_ARM_64 or something. We want one machine type per different image.

Reply:
:+1:

Reply:
If I push an updated `build_tools/docker/manage_images.py` as per my previous comment, will you build and store the arm docker images in gcp.io with the other docker images? I'm not experience enough with Github actions to dry code this change :smile:

Reply:
My assumption is; with the arm docker images on gcp.io, I can reference them in e.g. [build_test_runtime in ci.yml](https://github.com/openxla/iree/blob/6e499159b393846441cc28b30a6791dc46221ec4/.github/workflows/ci.yml#L350-L362).

Then, to actually run that test, presumably I need your help again to create the T2A instance using `build_tools/github_actions/runner/gcp/create_image.sh` . Not sure how I will be able to test changes to `create_image.sh`, but I can give the script a go. WDYT?

Reply:
I was looking into the doc and it seems like we don't have a process to update docker/images for people who have no GCP access yet. So here is a more concrete proposal for now:

1. Assume it is possible to test docker change on your side (e.g. set up your own t2a GCP VM, build docker images locally and run some tests), then you can work on the changes locally and send out for review once ready.
2. If there are any gcr.io references in changes, they can be left as TODO or using local docker image references.
3. We review the changes and approve it. Once the change is approved, IREE team members can help build and push docker images to IREE gcr.io and update the gcr.io references left as TODO in the previous step.
4. Then we can merge the PR.

Reply:
Yeah pushing new images requires permissions in the cloud project. And the way we have images reference each-other is by digest and for unfathomable reasons, the only way to get a digest is to push to a registry. So to use all the scripting and stuff, you need push permission. But under the hood this is all just docker. Especially if you're just rebuilding the base image, you can do that with normal docker commands or use the script but don't push (beware that that means that you won't have a digest to give to dependent images). Iterating locally on the builds shouldn't be too bad either: the docker_run.sh scripts (GitHub specific one and general docker one) just require a couple environment variables to tell them which directories to mount. If you have access to another docker registry, you could even push to that and then temporarily edit the CI file to point at your location. Or add a (again temporary) step to build the docker image on the CI and use the image locally. Anyway, I realize none of these are great. I think the ideal would be build steps to rebuild the docker images when they change, but you don't want to do that every time and you want something that people can still test locally.

Reply:
Thx @pzread @GMNGeoffrey .

> Assume it is possible to test docker change on your side (e.g. set up your own t2a GCP VM, build docker images locally and run some tests), then you can work on the changes locally and send out for review once ready.

Building the dockers and running the tests are no problem, I've done it on Arm AWS instances before pushing. Similarly, I can comment out the GCP stuff in `build_tools/docker/manage_images.py` and test that too I think. Now, I want to create an Arm job, something like `build_test_runtime_arm64` in `ci.yml`. On a first glance, I thought there was a bit to much duplication with `build_test_runtime`, so I wanted to break it out and parameterize it. My limited experience with Github actions made me think, "I should probably test is this first". This is when I felt the need to test it here on the upstream, sort of, "to tests the .yml changes". Makes sense?

But let me push an update to this PR, where I just duplicate `build_test_runtime` for Arm and leave the gcp.io references as TODO, as well as modifying `manage_images.py`. It easier to have some code to discuss around...

Reply:
I pushed a couple of patches to discuss around. The changes to `manage_images.py` can be improved since it need to be ran twice build to all necessary images. 1) Build all 2) Build `base` image for Arm. But it can serve as an example for now.

Reply:
@pzread @GMNGeoffrey @ScottTodd, another reflection from reading up on docker images. Would it be better to create a [multi-platform image](https://docs.docker.com/build/building/multi-platform/)? That way arm64/x86 can have the same gcr.io reference I guess? This is not my area of expertise, so I'd really appreciate some pointers how to proceed from here. Thx

Reply:
I was testing a little bit of multi-platform images with `docker buildx` and here is what I found.

First I tried to build a multi-platform image (x86_64 + arm64) and load it into the local docker cli (version 20.10.21):

```sh
docker pull ubuntu:20.04
# Get the digest of the multiarch image
# Update the `FROM ubuntu@sha256:<digest>`
docker buildx build --platform linux/amd64,linux/arm64 --file build_tools/docker/dockerfiles/base.Dockerfile --tag base:test . --load
```

It gave me the error:

```
ERROR: docker exporter does not currently support exporting manifest lists
```

From my reading (https://docs.docker.com/build/building/multi-platform/), a multi-platform image is a list of image manifests for each platform. It seems like currently the local docker cli doesn't support this format, which can be a problem for local development.

However, I think we want to use `docker buildx build --platform linux/arm64` to cross-build arm64 images (looks like it spawns qemu to do cross-build), so everyone can rebuild images with x86_64 machines. I'm also curious how you currently build the arm64 docker images in this PR?

The steps below can build an arm64 (single-platform) image on my x86_64 workstation:

```sh
docker buildx build --platform linux/arm64 --file build_tools/docker/dockerfiles/base.Dockerfile --tag base:arm64 --load .
docker image inspect base:arm64
```

Output:

```
[                                             
    {                                                                                                                                                                                                                                  
        "Id": "sha256:a12fbd04d2daffdf16d4bdca5c77daf5995fd66bf4e3e9d0240c506d2d077eba",                           
        "RepoTags": [                                                                                                                                                                                                                  
            "base:arm64"                                                                                                                                                                                                               
        ],                                                                                                                                                                                                                             
        "RepoDigests": [],                                                                                                                                                                                                             
...
```

Therefore, we might want to modify `manage_images.py` to use `docker buildx`.

Regarding the image organization, instead of having two SHA digests for `base@sha256:...` for two platforms, IMO we probably want to use different image names: `base` for x86_64 and `base-arm64` for arm64. This will make it easier and clearer to build other images based on them. For example, we can also have the `frontend` image based on `base` and `frontend-arm64` based on `base-arm64`.

There are several ways to implement that. One way is that `manage_images.py` has an option to specify which platform to build and it will add the platform as a suffix in image names. These will involve some efforts to make sure we can specify the supported platforms of each image in this list: https://github.com/openxla/iree/blob/main/build_tools/docker/manage_images.py#L44-L54

A simpler way for now is to keep `base.Dockerfile` for x86_64 and create `base-arm64.Dockerfile` for arm64. It looks like we can de-duplicate the code between two dockerfiles by factoring build steps into scripts (as what you did for `install_rocm.sh` and `install_vulkan.sh`). And some scripts don't need to run in `base-arm64.Dockerfile` (for example `install_rocm.sh`). This won't scale well when we start adding the arm64 variants for other docker images, but we can always move from here to the more complete solution above.

Reply:
Thx @pzread !

> I'm also curious how you currently build the arm64 docker images for this PR?

Currently, I build the arm64 dockers on an AWS Graviton (arm64) instances manually using `docker build ...`, or via the parameterized `manage_images.py` (but removing the registry interactions, since I don't have access). As per https://github.com/openxla/iree/issues/12684#issuecomment-1550413869, I make sure the `build_test_runtime` test works. Once we have this PR merged, I'll look into the rest of the relevant tests. 

> Regarding the image organization, instead of having two SHA digests for base@sha256:... for two platforms, IMO we probably want to use different image names: base for x86_64 and base-arm64 for arm64. This will make it easier and clearer to build other images based on them. For example, we can also have the frontend image based on base and frontend-arm64 based on base-arm64.

Let's try this approach, it'll make the image dependencies more modular as well. I'll push something shortly.

> However, I think we want to use docker buildx build --platform linux/arm64 in `manage_images.py` to cross-build arm64 images (looks like it spawns qemu to do cross-build), so everyone can rebuild images on x86_64 machines.

Ok, will do.

Reply:
@pzread I updated this PR with the changes we have discussed (force pushed to make it a bit cleaner). I tested `python3 build_tools/docker/manage_images.py --image base-arm64` to work on an x86 and arm64 machines, using my own registry.

If it looks ok, the next step is to upload the new image(s) to gcp.io. Then I guess I need help to create the GCP T2A runner as well.

Reply:
Oh right... I'll need some time to set up the T2A runner. I'll give you an update once it's done.

(FYI, I'll test image and runner here https://github.com/openxla/iree/pull/14444, because I'll need to push some test changes)

Reply:
Thx! I'll take care of the review comments